src/festim/coupled_heat_hydrogen_problem.py

```diff
@@ -0,0 +1,183 @@
+        hydrogen_problem: HydrogenTransportProblem,
+    ) -> None:
+        self.heat_problem = heat_problem
+        self.hydrogen_problem = hydrogen_problem
+
+        if (
+            not self.heat_problem.settings.transient
+            or not self.hydrogen_problem.settings.transient
+        ):
+            raise TypeError(
+                "Both the heat and hydrogen problems must be set to transient"
+            )
+
+    @property
+    def heat_problem(self):
+        return self._heat_problem
+
+    @heat_problem.setter
+    def heat_problem(self, value):
+        if not isinstance(value, HeatTransferProblem):
+            raise TypeError("heat_problem must be a festim.HeatTransferProblem object")
+        self._heat_problem = value
+
+    @property
+    def hydrogen_problem(self):
+        return self._hydrogen_problem
+
+    @hydrogen_problem.setter
+    def hydrogen_problem(self, value):
+        if isinstance(
+            value,
+            HTransportProblemDiscontinuous
+            | HTransportProblemPenalty
+            | HydrogenTransportProblemDiscontinuousChangeVar,
+        ):
+            raise NotImplementedError(
+                "Coupled heat transfer - hydorgen transport simulations with "
+                "HydrogenTransportProblemDiscontinuousChangeVar, "
+                "HTransportProblemPenalty or"
+                "HydrogenTransportProblemDiscontinuousChangeVar, "
+                "not currently supported"
+            )
+        elif not isinstance(value, HydrogenTransportProblem):
+            raise TypeError(
+                "hydrogen_problem must be a festim.HydrogenTransportProblem object"
+            )
+        self._hydrogen_problem = value
+
+    @property
+    def non_matching_meshes(self):
+        return self.heat_problem.mesh.mesh != self.hydrogen_problem.mesh.mesh
+
+    def initialise(self):
+        # make sure both problems have the same initial time step and final time,
+        # use minimal initial value of the two and maximal final time of the two
+        min_initial_dt = min(
+            self.heat_problem.settings.stepsize.initial_value,
+            self.hydrogen_problem.settings.stepsize.initial_value,
+        )
+        self.heat_problem.settings.stepsize.initial_value = min_initial_dt
+        self.hydrogen_problem.settings.stepsize.initial_value = min_initial_dt
+
+        if (
+            self.heat_problem.settings.final_time
+            != self.hydrogen_problem.settings.final_time
+        ):
+            raise ValueError(
+                "Final time values in the heat transfer and hydrogen transport "
+                "model must be the same"
+            )
+
+        self.heat_problem.initialise()
+
+        self.heat_problem.show_progress_bar = False
+
+        if self.non_matching_meshes:
+            V = fem.functionspace(self.hydrogen_problem.mesh.mesh, ("P", 1))
+            T_func = fem.Function(V)
+
+            nmm_interpolate(T_func, self.heat_problem.u)
```

Comment:
I guess we should be able to read the element family and order (P and 1) from `self.heat_problem` if we store this information

Reply:
Yes this is something we could do down the line, for now though these parameters aren't accessible to the users so may as well use the ones used in `HeatTransferProblem.define_function_space()`